native/src/main/java/io/ballerina/wso2/apim/catalog/ServiceCatalog.java

```diff
@@ -0,0 +1,263 @@
+    }
+
+    private static void updateServiceNameAndUrl(BMap<BString, Object> artifactValues,
+                                                HttpServiceConfig httpServiceConfig) {
+        updateServiceName(artifactValues, httpServiceConfig);
+        updateServiceUrl(artifactValues, httpServiceConfig);
+    }
+
+    private static void updateListenerConfigurations(BMap<BString, Object> artifactValues, Object listenerDetails) {
+        artifactValues.put(StringUtils.fromString(MUTUAL_SSL_ENABLED), getMutualSSLDetails(listenerDetails));
+    }
+
+    private static boolean getMutualSSLDetails(Object listenerDetails) {
+        ArrayList<BObject> listenerArray = (ArrayList<BObject>) listenerDetails;
+        if (!listenerArray.isEmpty()) {
+            BObject listener = listenerArray.get(0);
+            if (listener != null) {
+                Object configObject = listener.getNativeData(CONFIG);
+                if (configObject instanceof HashMap config
+                        && config.containsKey(StringUtils.fromString(SECURE_SOCKET))) {
+                    Object secureSocketObject = config.get(StringUtils.fromString(SECURE_SOCKET));
+                    return secureSocketObject instanceof HashMap secureSocket
+                            && secureSocket.containsKey(StringUtils.fromString(MUTUAL_SSL));
+                }
+            }
+        }
+        return false;
+    }
+
+    private static void updateServiceName(BMap<BString, Object> artifactValues, HttpServiceConfig httpServiceConfig) {
+        artifactValues.put(StringUtils.fromString(NAME),
+                StringUtils.fromString(httpServiceConfig.basePath));
+    }
+
+    private static void updateServiceUrl(BMap<BString, Object> artifactValues, HttpServiceConfig httpServiceConfig) {
+        String basePath = httpServiceConfig.basePath.equals(LOCALHOST) ? "" : httpServiceConfig.basePath;
+        artifactValues.put(StringUtils.fromString(SERVICE_URL),
+                StringUtils.fromString(new StringBuilder().append(httpServiceConfig.host).
+                        append(COLON).append(httpServiceConfig.port).append(basePath).toString()));
+        artifactValues.put(StringUtils.fromString(DEFINITION_URL),
+                StringUtils.fromString(httpServiceConfig.definitionUrl));
+    }
+
+    private static HttpServiceConfig updateHostAndPortAndBasePath(Object listenerDetails, Object attachPointDetails,
+                                                                  BMap<BString, Object> httpAnnotation) {
+        BArray attachPoints = (BArray) attachPointDetails;
+
+        String basePath = generateBasePath(attachPoints);
+        String port = getPortValue(listenerDetails);
+        String host = getHostname(httpAnnotation);
+        return new HttpServiceConfig(host, port, basePath);
+    }
+
+    private static String getPortValue(Object listenerDetails) {
+        ArrayList<BObject> listenerArray = (ArrayList<BObject>) listenerDetails;
```

Comment:
I think it is better to cast to the more general `List` instead of the concrete class

```suggestion
        List<BObject> listenerArray = (List<BObject>) listenerDetails;
```